packages/daimo-api/src/contract/keyRegistry.ts

```diff
@@ -0,0 +1,166 @@
+import { KeyData, contractFriendlyKeyToDER } from "@daimo/common";
+import { accountABI } from "@daimo/contract";
+import { Address, Hex, Log, getAbiItem, getAddress } from "viem";
+
+import { ViemClient } from "../chain";
+
+const signingKeyAddedEvent = getAbiItem({
+  abi: accountABI,
+  name: "SigningKeyAdded",
+});
+const signingKeyRemovedEvent = getAbiItem({
+  abi: accountABI,
+  name: "SigningKeyRemoved",
+});
+
+type SigningKeyAddedLog = Log<
+  bigint,
+  number,
+  false,
+  typeof signingKeyAddedEvent,
+  true
+>;
+type SigningKeyRemovedLog = Log<
+  bigint,
+  number,
+  false,
+  typeof signingKeyRemovedEvent,
+  true
+>;
+type SigningKeyAddedOrRemovedLog = SigningKeyAddedLog | SigningKeyRemovedLog;
+
+export class KeyRegistry {
+  /* In-memory indexer, for now. */
+  private keyToAddr = new Map<Hex, Address>();
+  private addrToLogs = new Map<Address, SigningKeyAddedOrRemovedLog[]>();
+  private addrToKeyData = new Map<Address, KeyData[]>();
+
+  constructor(private vc: ViemClient) {}
+
+  async init() {
+    await this.vc.pipeLogs(
+      {
+        event: signingKeyAddedEvent,
+      },
+      this.parseLogs
+    );
+
+    await this.vc.pipeLogs(
+      {
+        event: signingKeyRemovedEvent,
+      },
+      this.parseLogs
+    );
+    console.log(`[KEY-REG] watching logs`);
+  }
+
+  /** Get current keys of an address. */
+  async getKeys(addr: Address) {
+    try {
+      return await this.vc.publicClient.readContract({
+        abi: accountABI,
+        address: addr,
+        functionName: "getSigningKeys",
+      });
+    } catch (e: unknown) {
+      // Old Account, just ignore.
+      console.error(
+        `[API] Failed to get keys for ${addr}, probably an old account?`,
+        e
+      );
+      return [];
+    }
+  }
+
+  /** Parses account key add/remove logs, first on init() and then on subscription. */
+  parseLogs = async (logs: SigningKeyAddedOrRemovedLog[]) => {
+    const currentBlockNumber = await this.vc.publicClient.getBlockNumber(); // TODO: remove?
```

Comment:
i'd either remove this, or pass it into `getKeys` to ensure we're fetching keys as of that block number